### ISSUE TITLE
[Icons] design draft

### ISSUE THREAD

**Issue by mollieru** [18F/web-design-standards#43]
For iconography, we have been considering forking the way that USPTO has organized their reuse of Font-Awesome, which we're also suggesting we use. Would be helpful to hear how others feel about this idea. Screenshot below and here's the live link: https://uspto.github.io/designpatterns/1.x/docs/foundation/icons.html ]

Leaving this here for feedback:
![icons uspto ui design library](https://cloud.githubusercontent.com/assets/3680494/8219678/b2a58616-1519-11e5-8da9-1c48a8b107d6.png)

**Comment by arvalarva** [comment]
So far, Font Awesome has met the vast majority of the U.S. Patent and Trademark Office's needs. The community can also request new icons, or simply add icons within GitHub to the core Font Awesome library. Also, Font Awesome is becoming so widespread, that the goal of easily recognizable icons (instructions) for our customers is easily met.

**Comment by logantpowell** [comment]
:+1:  "God is in the details" - Mies Van der Rohe

**Comment by carodew** [comment]
We've heard some concerns in user interviews that Font Awesome can be a bit cartoony, but it's not been widespread feedback. I do think that it would be worthwhile at some point to remove some of the cheesier icon options (beer? Checkered flag?) and/or recommend specific icons for specific contexts.

**Comment by logantpowell** [comment]
:-1: "Beer is the proof that god loves us and wants us to be happy" - Benjamin Franklin

**Comment by carodew** [comment]
@logantpowell touché.

**Comment by andrewmaier** [comment]
Icon fonts are _almost_ an anti-pattern for most applications: they're difficult to update/maintain, not semantic, and poorly rendered/supported. Instead, I'd recommend we use SVGs with a PNG fallback. There's a really good presentation on the benefits of this approach, here: https://docs.google.com/presentation/d/1CNQLbqC0krocy_fZrM5fZ-YmQ2JgEADRh3qR6RbOOGk/edit#slide=id.p

**Comment by logantpowell** [comment]
I'm not an expert, but I'd say that - if the intention is to create an icon set for everyone to share, this is a good way of doing it. I.e., in terms of maintenance/updates/etc... However, you have a good point about the support. 

I would say it depends on the market you're thinking of providing a solution for... Here's a [take](https://clicky.com/marketshare/us/web-browsers/mobile/) on the US mobile browser market:

![mobile_web_browsers_ us_marketshare _clicky_-_2015-06-22_16 32 01](https://cloud.githubusercontent.com/assets/2817224/8292207/512cf75c-18fc-11e5-9906-d58fff9f8d33.png)

**Comment by maya** [comment]
@andrewmaier doesn't Font Awesome also provide SVGs? Are we just talking about what implementation we want?

**Comment by mjovel** [comment]
@andrewmaier @maya Normally I would say I would recommend a SVG based system but as this is an icon set which in theory would be shared across multiple agencies I feel like font based solution like Font Awesome offers the most efficient method of loading icons. with SVG w/ Png fallback we would either have to worry about multiple http requests or explaining automation of spritesheet creation, or a larger generic spritesheet with icons that are not being used. Do we have metrics from DAP that estimate % of gov site visitors that use devices that do not support @font-face?

**Comment by logantpowell** [comment]
Could perhaps some best practices for making these semantic be included in the spec? Like telling folks to include a label of the meaning (e.g., "email icon") of the icon? Idk... I figure someone has some ideas about this...

**Comment by arvalarva** [comment]
Be careful in selecting a single solution. Some large firms and government agencies have security settings that block IE from using webfonts. As you know, we are using Font Awesome, but have an SVG option, since we are one of those government agencies here at the USPTO.

**Comment by mjovel** [comment]
@arvalarva Good call on the webfont blocking! Is there a way to efficiently offer more than one solution? Would it be better to just take the ~450k hit for the full SVG/png spritesheets and leave it to the advanced users to trim that down via for individual implementations?

**Comment by andrewmaier** [comment]
@mjovel The script referenced in the presentation I suggested handles this efficiently by only loading the most contemporary assets supported by a user's browsing agent. You can see this programmatically, here:

https://github.com/filamentgroup/grunticon/blob/master/example/output/preview.html

Specifically this bit:

`grunticon(["icons.data.svg.css", "icons.data.png.css", "icons.fallback.css"], grunticon.svgLoadedCallback);`

It's essentially loading SVGs if they're supported, PNGs if they're not, and then a CSS sprite in the "worst" case.

**Comment by mjovel** [comment]
@andrewmaier I agree that from a technology standpoint svg with fallbacks and even more so grunticons is the best way to go.  I have used both on a few projects (not gov stuff).  That said my concerns is for a framework that is supposed to be used across many agencies with many stacks and with people of many technical abilities if it is too tall an order? Truthfully, this is not a question I can answer. To be honest I have had quite a bit of trouble on getting my agency onboard with moving off of webfont icons which is where my concern stems from.

**Comment by andrewmaier** [comment]
@mjovel I 1000% agree with you there. I think, though, that the choice is that we either promulgate sub-par practices—say, icon fonts—or we point the way towards the "best" ones as of the time we create the pattern library.

**Comment by logantpowell** [comment]
Could we do something like [this](http://code.tutsplus.com/tutorials/quick-tip-ever-thought-about-using-font-face-for-icons--net-11233)

**Comment by arvalarva** [comment]
@Font-face doesn’t work in IE behind our firewall. Web-based font icons work in Chrome and Firefox, even behind our firewall. We tested, and some of our customers at large firms also cannot see font-based icons (or any web fonts, for that matter) using IE behind their firewalls.

It’s a great solution, but you need a properly coded svg back-up, as I saw someone else suggest in this thread. Right? I’m adding Brian, who knows more.

Thanks,
Arva

**Comment by mjovel** [comment]
@andrewmaier My concern is the creation of the dependency of teams having to run a grunt task (or do something like grumpicons) to reduce amount of data downloaded either via http request or uri data in css to display icons. I fear implementing a workflow change like that can be dealbreaker for some. I am kind of curious what the size of the icons-data-svg.css and icons-data-png-css files for the full ustpo/awesomefont library of icons?

**Comment by bmathews** [comment]
@mjovel: The USPTO set is 115k for the gzipped svg sprite and 5k for the gzipped icon CSS. Does that answer your question, or did I misunderstand? Inlining all the svgs into the css would probably be about the same size. Font Awesome is pretty big, and there are probably quite a few icons that'll never get used by the USPTO.

It looks like you've all discussed pretty much all the options, and they all have support/color/flexibility/maintenance/size pros and cons to weigh. It's not a fun decision to make, and it does sometime depend on the tech stack, tooling, and team like @mjovel mentioned. 

At the very least, maintaining and providing all the source svgs individually gives your consumers some flexibility if they aren't able to use whatever method you end up going with. 

[Detecting whether IE is configured to block webfont downloads](http://stackoverflow.com/questions/7973421/detect-if-font-download-in-ie-is-enabled) isn't great.

**Comment by mollieru** [comment]
Comment from @mariamarrero in issue https://github.com/18F/govt-wide-patternlibrary/issues/125 :
"Using web app icons sounds like a good option. However, agencies may have the need for icons that are not included in that list. In that case, they may be forced to create their own icons. Perhaps we can also mention some specs for the icons if there is that need?

Also, some agencies may have some security guidelines, block them and may not be able to use them."

**Comment by mollieru** [comment]
Comment from @afeijoo in issue https://github.com/18F/govt-wide-patternlibrary/issues/125 :

"+1 @mariamarrero. We've found that quite a few agencies block Font Awesome icons, especially those with older versions of Internet Explorer."

**Comment by mollieru** [comment]
Comment from @jjoteal in:

"https://pages.18f.gov/govt-wide-patternlibrary/visual-style/#icons
Specify source for icons (Font Awesome, etc.)"

**Comment by carodew** [comment]
Moving this issue into our team's workflow to consider all the feedback. Thanks!

**Comment by maya** [comment]
We will be moving forward with SVG with PNG fallback due to: 

1. Government agencies that have security settings blocking IE from using webfonts
2. Basic accessibility issues with the lack of alt text on icon fonts

Thanks all for this lively and informative discussion!

**Comment by JJediny** [comment]
Never appropriately licensed these all from all the sources but I'd previously collected all the open licensed svg/ttf repos out their for cartography purposes but this may give others ideas and places to look:
https://github.com/JJediny/NEPAnode-MapIcons

**Comment by bendygirl** [comment]
The government of Buenos Aires created fonts for use across all of BA. They're available here http://gcba.github.io/iconos/ best use of icons (in my opinion) their last one, the Tango. Formats are EPS and PNG

**Comment by jhfrench** [comment]
Does the [IEOld end-of-life](https://www.microsoft.com/en-us/WindowsForBusiness/End-of-IE-support) (moment of silence, as we contemplate the terrors that once visited upon our code), and subsequent mandate to adopt IE11, change any of the logic against using webfonts?

**Comment by jhfrench** [comment]
> Basic accessibility issues with the lack of alt text on icon fonts


@maya I believe [alt text is required/expected for meaningful images](https://www.w3.org/WAI/tutorials/images/decorative/), particularly when the meaning of that image is not otherwise conveyed. So if you have...

```
<i class="fa fa-refresh fa-spin fa-3x fa-fw" aria-hidden="true"></i>
<span class="sr-only">Refreshing...</span>
```

...than you're good:

- The sighted user sees a spinning refresh icon, but no "Refreshing..." text
- A non-sighted user gets "Refreshing...", but is otherwise unecumbered by a trivial image (which is not even really an `img`, so it doesn't need an `alt` attribute, nor does it get "read" to the user!)

If I'm wrong about this, someone please re-educate me quickly!

**Comment by Drakenhart** [comment]
@jhfrench First, the idea seems doable but would need to be tested to see if it works well for screen-readers.

An added point - if it is an alert status of some kind you will have to make sure screen readers users understand it. There needs to be context. Ex: If an alert is informational, it should state that. If it is a warning, it should state that.

**Comment by jbrucegsa** [comment]
Another alternative would be to include both an icon and text. This would be a more "mobile-first" approach as there is no "hover" state on mobile; therefore, no display of the alternative text as a tool tip.

One of the things we are attempting at my program:

<img width="204" alt="screen shot 2016-04-05 at 9 24 38 am" src="https://cloud.githubusercontent.com/assets/15252830/14691180/d1f34660-071d-11e6-9a52-05d9aacc8a0f.png">
<img width="139" alt="screen shot 2016-04-05 at 9 24 43 am" src="https://cloud.githubusercontent.com/assets/15252830/14691184/d5f4531c-071d-11e6-8ff2-e742f5933a7e.png">

This way the icon is not the *only* communication vehicle - this is particularly helpful when the icon is not as obvious to communicate ideas. For example, what is the appropriate icon for an "unarchive" action?

**Comment by jhfrench** [comment]
@jbrucegsa  

> ...what is the appropriate icon for "unarchive" action?

Clearly @davegandy needs to devise an icon that conveys [this sentiment](https://youtu.be/1DFojKwVfq4?t=24s)

**Comment by Drakenhart** [comment]
@jbrucegsa That is what my project is doing. We limit our icons use for one to only items that are either very important, or need to be quickly found through users scanning the site. We pair them up with appropriate text as well. If we feel any additional information is needed for screen readers but might interfere with the visual design, we use a SR-only class to hide the information visually.

As for "unarchive" why just use" move" instead? :)

@jhfrench epic, undead epic.

@maya The CMS we use has a folder for the web fonts and the web icons, so if someone can reach the web service we can still serve the fonts and icons from the on-network server.

I'm not sure if that is a good solution or not, because I do believe that in some cases the block may be local to the computer itself?

**Comment by jhfrench** [comment]
@bendygirl I just need everyone to know that I *want* to live in a world where the government of Buenos Aires created an icon project code-named "Tango"

This post may or may not have added a meaningful contribution to the conversation :grin:

**Comment by jpdevries** [comment]
> I feel like font based solution like Font Awesome offers the most efficient method of loading icons

Well usually with Font Awesome you are loading 600+ icons even if you need to use two. Yes you can use IcoMoon to create a web font with just what you need, but not many people do. I think an SVG implementation and/or SVG `<use>` would be the most efficient. 

FontAwesome is moving towards supporting SVG icons natively! But there are ways to use build tools like Grunt to create your own SVG sprites of SVG icons (Font Awesome or any other) today
https://markup.tips/projects/grunting-icons-into-svg-sprites.html

**Comment by joshbruce** [comment]
@jhfrench - Sorry it's taken me a moment. For unarchive (undo), I recommend, the FA undo icon: http://fontawesome.io/icon/undo/

**Comment by joshbruce** [comment]
@jpdevries et. al. - You can also include the library and fonts, which overcomes the limitation of seeking outside/CDN resources for those "more secure" locations. You lose the ability to take advantage of a cache from a CDN, but you can also be a bit more picky by optimizing the FA Sass if you want to do that sort of thing. I think it would be a later performance optimization in the grand scheme of things.

The USWDS is already doing this with the inclusion of the Source Sans and Merriwhether (sp?) fonts.

**Comment by ostranme** [comment]
Have you considered govicons (http://govicons.io/)? I've tried to align this with fontawesome as much as possible but making it specific to government themed icons.

**Comment by yowill** [comment]
CFPB has a set too that's neato: https://cfpb.github.io/design-manual/brand-guidelines/minicons

**Comment by jpdevries** [comment]
I've been doing some research on SVG icons vs icon fonts in https://github.com/tootsuite/mastodon/issues/1469#issue-220749234

There's some browser stats and Pros and Cons there. 

**TL;DR**
 - SVG has better browser support than font face
 - SVG is more accessible for user agents that may change the font(s) of the page

**Comment by shawnbot** [comment]
I'm solidly in the SVG camp. Here are a few good write-ups to consider:

* [Inline SVG vs. icon fonts [CAGEMATCH]](https://css-tricks.com/icon-fonts-vs-svg/)
* [Seriously, don't use icon fonts](https://cloudfour.com/thinks/seriously-dont-use-icon-fonts/)
* [Ten reasons we [Lonely Planet] switched from an icon font to SVG](http://ianfeather.co.uk/ten-reasons-we-switched-from-an-icon-font-to-svg/)

**Comment by maya** [comment]
Thanks for a great discussion all. Moving future icons development to issue #762.
